app/src/main/java/org/wikipedia/diff/ArticleEditDetailsFragment.kt

```diff
@@ -0,0 +1,415 @@
+                .getStringExtra(EXTRA_SOURCE_ARTICLE_TITLE), "")
+        revisionId = requireActivity().intent
+                .getLongExtra(EXTRA_SOURCE_EDIT_REVISION_ID, 0)
+        languageCode = StringUtils.defaultString(requireActivity().intent
+                .getStringExtra(EXTRA_SOURCE_EDIT_LANGUAGE_CODE), "en")
+
+        setUpInitialUI()
+        setUpListeners()
+        fetchEditDetails()
+    }
+
+    private fun setUpListeners() {
+        articleTitleView.setOnClickListener {
+            val title = PageTitle(articleTitle, WikiSite.forLanguageCode(languageCode))
+            if (title.namespace() == Namespace.USER_TALK || title.namespace() == Namespace.TALK) {
+                startActivity(newIntent(requireContext(), title.pageTitleForTalkPage()))
+            } else {
+                startActivity(PageActivity.newIntentForNewTab(requireContext(),
+                        HistoryEntry(title, HistoryEntry.SOURCE_ON_THIS_DAY_ACTIVITY), title))
+            }
+        }
+        newerIdButton.setOnClickListener {
+            revisionId = newerRevisionId
+            fetchEditDetails()
+        }
+        olderIdButton.setOnClickListener {
+            revisionId = olderRevisionId
+            fetchEditDetails()
+        }
+        watchButton.setOnClickListener {
+            watchOrUnwatchTitle(watchlistExpirySession, isWatched)
+        }
+        usernameButton.setOnClickListener {
+            if (AccountUtil.isLoggedIn() && username != null) {
+                startActivity(newIntent(requireActivity(),
+                        PageTitle(UserTalkAliasData.valueFor(languageCode),
+                                username!!, WikiSite.forLanguageCode(languageCode))))
+            }
+        }
+        thankButton.setOnClickListener { showThankDialog() }
+    }
+
+    private fun setUpInitialUI() {
+        diffSize = requireActivity().intent.getIntExtra(EXTRA_SOURCE_EDIT_SIZE, 0)
+        diffText.movementMethod = ScrollingMovementMethod()
+        articleTitleView.text = articleTitle
+        if (diffSize >= 0) {
+            diffCharacterCountView.setTextColor(if (diffSize > 0) ContextCompat.getColor(requireContext(),
+                    R.color.green50) else ResourceUtil.getThemedColor(requireContext(), R.attr.material_theme_secondary_color))
+            diffCharacterCountView.text = String.format("%+d", diffSize)
+        } else {
+            diffCharacterCountView.setTextColor(ContextCompat.getColor(requireContext(), R.color.red50))
+            diffCharacterCountView.text = String.format("%+d", diffSize)
+        }
+    }
+
+    private fun fetchEditDetails() {
+        disposables.add(Observable.zip(ServiceFactory.get(WikiSite.forLanguageCode(languageCode)).getRevisionDetails(articleTitle, revisionId),
+                ServiceFactory.get(WikiSite.forLanguageCode(languageCode)).getWatchedInfo(articleTitle),
+                { r, w ->
+                    isWatched = w.query()!!.firstPage()!!.isWatched
```

Comment:
Would it be possible to add checks to check the `query()` and `firstPage()` to prevent untraceable errors instead of using many `!!`?